### PR TITLE
oc: wire set image from kubectl

### PR DESCRIPTION
**Pull request by kargakis**
Wiring kubectl set image which came with the rebase

@mfojtik ptal

cc: @openshift/cli-review @smarterclayton @tnozicka

### CODE REVIEW

File: pkg/cmd/cli/cmd/set/set.go [hunk 1]
```diff
@@ -57,3 +59,29 @@ func NewCmdSet(fullName string, f *clientcmd.Factory, in io.Reader, out, errout
...
+const (
+	setImageLong = `
+Update existing container image(s) of resources.
+`
+
+	setImageExample = `	# Set a deployment configs's nginx container image to 'nginx:1.9.1', and its busybox container image to 'busybox'.
+	%[1]s image dc/nginx busybox=busybox nginx=nginx:1.9.1
+
+	# Update all deployments' and rc's nginx container's image to 'nginx:1.9.1'
+	%[1]s image deployments,rc nginx=nginx:1.9.1 --all
```
Comment (mfojtik) [inline pkg/cmd/cli/cmd/set/set.go]: this is misleading right? it should be `dc` not deployments or it should be `rs` instead of `rc` ;-)

Comment (mfojtik) [inline pkg/cmd/cli/cmd/set/set.go]: although I don't know if updating the image for deployment config is a good idea (if it have the ICT)

File: pkg/cmd/cli/cmd/set/set.go [hunk 2]
```diff
@@ -57,3 +59,29 @@ func NewCmdSet(fullName string, f *clientcmd.Factory, in io.Reader, out, errout
...
+
+	# Update image of all containers of daemonset abc to 'nginx:1.9.1'
+	%[1]s image daemonset abc *=nginx:1.9.1
+
+	# Print result (in yaml format) of updating nginx container image from local file, without hitting the server 
+	%[1]s image -f path/to/file.yaml nginx=nginx:1.9.1 --local -o yaml`
```
Comment (mfojtik) [inline pkg/cmd/cli/cmd/set/set.go]: i wonder why or when we diverged from the `--dry-run`

File: pkg/cmd/cli/cmd/set/set.go [hunk 3]
```diff
@@ -57,3 +59,29 @@ func NewCmdSet(fullName string, f *clientcmd.Factory, in io.Reader, out, errout
...
+)
+
+// NewCmdImage is a wrapper for the Kubernetes cli set image command
```
Comment (mfojtik) [inline pkg/cmd/cli/cmd/set/set.go]: nit: `Kubernetes CLI set image`

File: test/cmd/deployments.sh [hunk 4]
```diff
@@ -132,6 +132,14 @@ os::cmd::expect_success 'oc delete hpa/test-deployment-config'
 echo "autoscale: ok"
 os::test::junit::declare_suite_end
 
+os::test::junit::declare_suite_start "cmd/deployments/setimage"
+os::cmd::expect_success 'oc create -f test/integration/testdata/test-deployment-config.yaml'
+os::cmd::expect_success 'oc set image dc/test-deployment-config ruby-helloworld=myshinynewimage'
+os::cmd::expect_success_and_text "oc get dc/test-deployment-config -o jsonpath='{.spec.template.spec.containers[0].image}'" "myshinynewimage"
+os::cmd::expect_success 'oc delete dc/test-deployment-config'
```
Comment (mfojtik) [inline test/cmd/deployments.sh]: can we also have negative test? (iow. setting image for build config should return a reasonable error)... i'm ok with some follow up.

File: pkg/cmd/cli/cmd/set/set.go (hunk 1 above)
Comment (kargakis) [inline pkg/cmd/cli/cmd/set/set.go]: It's just an example of using set image with multiple resources, it's not scoped specifically to DC+RC or D+RS

File: test/cmd/deployments.sh (hunk 4 above)
Comment (kargakis) [inline test/cmd/deployments.sh]: Will spawn an issue since this will probably require upstream changes

Comment (kargakis) [inline test/cmd/deployments.sh]: https://github.com/openshift/origin/issues/10947

**Comment by kargakis** [comment]
[test]

**Comment by mfojtik** [comment]
LGTM

File: test/cmd/deployments.sh (hunk 4 above)
Comment (bparees) [inline test/cmd/deployments.sh]: how does this "just work" for deployment configs?  presumably the upstream implementation doesn't understand DCs...

Comment (mfojtik) [inline test/cmd/deployments.sh]: @bparees it is looking up for pod spec, like `oc set env`

**Comment by kargakis** [comment]
exactly

File: test/cmd/deployments.sh (hunk 4 above)
Comment (bparees) [inline test/cmd/deployments.sh]: @mfojtik so it just looks for a pod spec within any arbitrary resource it is given?

**Comment by mfojtik** [comment]
yes

-- 
Michal Fojtik

File: pkg/cmd/cli/cmd/set/set.go [hunk 5]
```diff
@@ -57,3 +59,29 @@ func NewCmdSet(fullName string, f *clientcmd.Factory, in io.Reader, out, errout
 	templates.ActsAsRootCommand(set, []string{"options"}, groups...)
 	return set
 }
+
+const (
+	setImageLong = `
+Update existing container image(s) of resources.
+`
+
+	setImageExample = `	# Set a deployment configs's nginx container image to 'nginx:1.9.1', and its busybox container image to 'busybox'.
```
Comment (fabianofranz) [inline pkg/cmd/cli/cmd/set/set.go]: This seems a tab, use 2 spaces in examples. BTW we upstreamed our templating system with some improvements including correcting examples tabbing automatically, adding a `post-rebase` issue to myself to reuse templating from Kube. Once I do that we'll no longer need to worry about this.

**Comment by kargakis** [comment]
done

**Comment by smarterclayton** [comment]
Set image needs to accept image stream tags and image stream images.

**Comment by mfojtik** [comment]
@smarterclayton and resolve them? a follow up to me.

**Comment by kargakis** [comment]
> Set image needs to accept image stream tags and image stream images.

Opened https://github.com/openshift/origin/issues/10974

**Comment by smarterclayton** [comment]
If we're not going to fix it in 1.4 I have serious concerns about exposing
this.  It makes the experience problematic with our registry.  What are the
odds someone on your team is able to implement that for 1.4 (probably
relevant for oc run as well)?

On Sep 17, 2016, at 1:28 PM, Michail Kargakis <notifications@github.com>
wrote:

Set image needs to accept image stream tags and image stream images.

Opened #10974 <https://github.com/openshift/origin/issues/10974>

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/10945#issuecomment-247791718>,
or mute
the thread

**Comment by mfojtik** [comment]
I can look at that on Monday.

-- 
Michal Fojtik

**Comment by kargakis** [comment]
@mfojtik started working on the follow-up so [merge]

**Comment by mfojtik** [comment]
ship it!

**Pull request merged**